### PR TITLE
replaced box ray intersection

### PR DESCRIPTION
**Pull request by LeroyR**
https://github.com/ros-planning/geometric_shapes/issues/68

### CODE REVIEW

**Comment by davetcoleman** [comment]
This is confusing how you just have a single number for the description, can you please elaborate?

**Comment by LeroyR** [comment]
#68 is a link to the issue i opened 6 month ago.

`https://github.com/ros-planning/geometric_shapes/issues/68`

**Comment by davetcoleman** [comment]
Should https://github.com/ros-planning/geometric_shapes/pull/67 be merged after this PR?

File: src/bodies.cpp [hunk 1]
```diff
@@ -579,76 +579,72 @@ void bodies::Box::computeBoundingCylinder(BoundingCylinder &cylinder) const
...
-  if (detail::distanceSQR(center_, origin, dir) > radius2_) return false;
-
-  double t_near = -std::numeric_limits<double>::infinity();
-  double t_far  = std::numeric_limits<double>::infinity();
 
-  for (int i = 0; i < 3; i++)
+  //Brian Smits. Efficient bounding box intersection. Ray tracing news 15(1), 2002
+  float tmin, tmax, tymin, tymax, tzmin, tzmax;
+  float divx, divy, divz;
+  divx = 1 / dir.x();
```
Comment (davetcoleman) [inline src/bodies.cpp:586]: Check that dir.x() is not zero

File: src/bodies.cpp [hunk 2]
```diff
@@ -579,76 +579,72 @@ void bodies::Box::computeBoundingCylinder(BoundingCylinder &cylinder) const
 
 bool bodies::Box::intersectsRay(const Eigen::Vector3d& origin, const Eigen::Vector3d& dir, EigenSTL::vector_Vector3d *intersections, unsigned int count) const
```
Comment (davetcoleman) [inline src/bodies.cpp:580]: line width too long
side note: We should add clang-format checking to this repo

File: src/bodies.cpp [hunk 3]
```diff
@@ -579,76 +579,72 @@ void bodies::Box::computeBoundingCylinder(BoundingCylinder &cylinder) const
...
+    tmax = (corner2_.x() - origin.x()) * divx;
+  } else
+  {
+    tmax = (corner1_.x() - origin.x()) * divx;
+    tmin = (corner2_.x() - origin.x()) * divx;
+  }
 
-      if (t1 > t2)
-        std::swap(t1, t2);
+  divy = 1 / dir.y();
```
Comment (davetcoleman) [inline src/bodies.cpp:597]: divide by zero

**Review by davetcoleman — CHANGES_REQUESTED** [review]

**Comment by LeroyR** [comment]
The division by zero is needed here, division of a finite nonzero float by 0 is well-defined and results in +infinity (if the value was >zero) or -infinity (if the value was less than zero). 

#67 can be merged after this but the tests are just random failed cases. Best case would be to investigate why the rays did not work with the used algorithm. 

I could clang-format this but there are many long lines in this file alone

**Comment by rhaschke** [comment]
@LeroyR Please rebase this PR on top of #67. This way we can merge the breaking test and the fix together.

File: test/test_point_inclusion.cpp [hunk 4]
```diff
@@ -211,6 +211,50 @@ TEST(BoxRayIntersection, SimpleRay1)
...
+
+    intersect = box->intersectsRay(ray_o, ray_d.normalized(), &p);
+    EXPECT_TRUE(intersect);
+
+    delete box;
+}
+
+TEST(BoxRayIntersection, BrokenRay2)
+{
+    shapes::Box shape(0.02, 0.40000000000000002, 1.2);
```
Comment (rhaschke) [inline test/test_point_inclusion.cpp]: Please, clean up those floats.

File: test/test_point_inclusion.cpp [hunk 5]
```diff
@@ -211,6 +211,50 @@ TEST(BoxRayIntersection, SimpleRay1)
     EXPECT_TRUE(intersect);
 }
 
+TEST(BoxRayIntersection, BrokenRay)
```
Comment (rhaschke) [inline test/test_point_inclusion.cpp]: Can you find more descriptive names for the unittests?

**Review by rhaschke — CHANGES_REQUESTED** [review]
+1
Some minor changes: Please cleanup the unit tests.
Do you know whether other shape primitives have similar issues?

File: test/test_point_inclusion.cpp (hunk 5 above)
Comment (LeroyR) [reply test/test_point_inclusion.cpp]: I dont know what caused the failures, so any suggestions? NonSimpleRay1&2 ?
> the tests are just random failed cases. Best case would be to investigate why the rays did not work

**Comment by LeroyR** [comment]
> Do you know whether other shape primitives have similar issues?

No, but could be this was a pretty random discovery as i sample a good amount of rays. I may add other shapes in the future but this was a pretty random discovery.

File: test/test_point_inclusion.cpp (hunk 5 above)
Comment (rhaschke) [reply test/test_point_inclusion.cpp]: What about SimpleRay2&3 then?

**Review by rhaschke — APPROVED** [review]

File: src/bodies.cpp (hunk 2 above)
Comment (LeroyR) [reply src/bodies.cpp:580]: no changes by me

File: src/bodies.cpp (hunk 1 above)
Comment (LeroyR) [reply src/bodies.cpp:586]: The division by possible zero is needed here

File: src/bodies.cpp (hunk 3 above)
Comment (LeroyR) [reply src/bodies.cpp:597]: division of a finite nonzero float by 0 is well-defined and results in +infinity (if the value was >zero) or -infinity (if the value was less than zero).

**Pull request merged**